### ISSUE TITLE
porting to keras

### ISSUE THREAD

**Issue by michaelosthege** [tkipf/gcn#5]
Hi Thomas,
first of all thank you for sharing this implementation!
I intend to port this to keras and before running into a dead end, I would like to ask: Can you think of any design differences that would it make impossible to do this?
cheers
michael

**Comment by tkipf** [comment]
Hi Michael,

Thanks for your question. Porting to keras is indeed possible. I have been internally working with this for some time now, but I'm happy to share this with the community. You can find the code here:

https://github.com/tkipf/kegra

Contributions are welcome!

Best,
Thomas

**Comment by tkipf** [comment]
Note that this keras port is still in a very early stage and mostly undocumented. Contributions are welcome!

**Comment by michaelosthege** [comment]
With a bit of hacking, keras can be tweaked to `feed_dict` 3D sparse placeholders, but I ran into a dead end just before the finish line: [Tensorflow does not implement sparse*dense multiplication with >2D sparse tensors](https://www.tensorflow.org/versions/master/api_docs/python/sparse_ops/math_operations#sparse_tensor_dense_matmul).
This is not a problem in your original implementation, because every training step just takes one example. In keras there's currently no support for non-batched inputs and even batchsize 1 does add a third dimension.

For the other things I copied keras.layers.Dense and modified it with the adjacency-multiplication. The loss function is more tricky: The mask has to be fed as an additional input to the model. The loss function can then be defined like this:

```python
def create_masked_accuracy(mask):
    """Accuracy with masking."""
    def masked_accuracy(y_true, y_pred):
        correct_prediction = K.equal(K.argmax(y_pred, 1), K.argmax(y_true, 1))
        accuracy_all = K.cast(correct_prediction, "float32")
        accuracy_all *= prepare_mask(mask) # normalize the mask
        return K.mean(accuracy_all)
    return masked_accuracy
```

I ran out of ideas for today..

**Comment by tkipf** [comment]
This is indeed a problem that I ran into at some point as well.

As a workaround, you can take the number of nodes in the graph as your `batch_size` and feed the feature matrix effectively as a `batch_size x num_features` (sparse) 2D matrix and the adjacency matrix as a `batch_size x batch_size` sparse 2D matrix (see  [tkipf/kegra](https://github.com/tkipf/kegra) for an example). This poses some other problems as the `batch_size` is deeply interwoven in keras and keras expects the output of your model to be of shape `batch_size x (other dimensions)` which can be limiting in case you want to do graph-level classification or something like the [Variational Graph Auto-Encoder](http://bayesiandeeplearning.org/papers/BDL_16.pdf).

**Comment by michaelosthege** [comment]
I was able to modify keras to accommodate for single-instance batches that have one dimension less, but
1. it does not converge (accuracy goes up, then back down)
2. sample_weights are embedded even deeper and very hard to remove
I would like to do graph-level classification, but it seems to be almost impossible to make keras work with single instances instead of batches.

**Comment by michaelosthege** [comment]
I've managed to get all the things going.

Considering that there's the kegra repository, I'll leave a few notes here and close the issue:

+ keras sparse placeholders aren't initialized with proper shapes, sometimes leading to unexpected behaviors with metrics. As a workaround the product of sparse & dense tensors can be reshaped to restore proper shapes.
+ single-instance training gives more flexibility and it's a fun exercise to implement

cheers
